ethereum/rpc/namespaces/eth/filters/filters.go

```diff
@@ -207,21 +186,52 @@ func (f *Filter) blockLogs(header *ethtypes.Header) ([]*ethtypes.Log, error) {
 	return logs, nil
 }
 
-// checkMatches checks if the logs from the a list of transactions transaction
-// contain any log events that  match the filter criteria. This function is
-// called when the bloom filter signals a potential match.
-func (f *Filter) checkMatches(transactions []common.Hash) []*ethtypes.Log {
-	unfiltered := []*ethtypes.Log{}
-	for _, tx := range transactions {
-		logs, err := f.backend.GetTransactionLogs(tx)
-		if err != nil {
-			// ignore error if transaction didn't set any logs (eg: when tx type is not
-			// MsgEthereumTx or MsgEthermint)
+func createBloomFilters(filters [][][]byte) [][]BloomIV {
+	bloomFilters := make([][]BloomIV, 0)
+	for _, filter := range filters {
+		// Gather the bit indexes of the filter rule, special casing the nil filter
+		if len(filter) == 0 {
 			continue
 		}
+		bloomIVs := make([]BloomIV, len(filter))
+		for i, clause := range filter {
+			if clause == nil {
+				bloomIVs = nil
+				break
+			}
+			bloomIVs[i] = calcBloomIVs(clause)
+		}
+		// Accumulate the filter rules if no nil rule was within
+		if bloomIVs != nil {
+			bloomFilters = append(bloomFilters, bloomIVs)
+		}
+	}
+	return bloomFilters
+}
 
-		unfiltered = append(unfiltered, logs...)
+// calcBloomIVs returns BloomIV for the given data,
+// revised from https://github.com/ethereum/go-ethereum/blob/401354976bb44f0ad4455ca1e0b5c0dc31d9a5f5/core/types/bloom9.go#L139
+func calcBloomIVs(data []byte) BloomIV {
+	hashbuf := make([]byte, 6)
+	biv := BloomIV{}
+
+	sha := crypto.NewKeccakState()
+	sha.Reset()
+	if _, err := sha.Write(data); err != nil {
+		panic(err)
 	}
+	if _, err := sha.Read(hashbuf); err != nil {
+		panic(err)
```

Comment:
why panic? should we validate the data field to prevent panics?

Reply:
I think the error in `sha.Read` should not happens, in the same function of the go-ethereum project, it doesn't check the err in the `sha.write` and `sha.read`. I put the check because of the lint error.

But we can limit the input size in case something happens.

Reply:
return error to instead of the panic.
So if the error happens, the filter rules will be skipped.